### PR TITLE
Use EVP interface for ECDH in apps/speed.c

### PR DESCRIPTION
**Pull request by romen**
Noticed [this thread in the openssl-dev ML](https://mta.openssl.org/pipermail/openssl-dev/2016-October/008582.html).

I was already working on option number 3 for a side project where I had the need to compare benchmarks of ECDH with different curves, including X25519, so here is a pull request to start from if we want to revise which interface to use to access EC crypto in apps/speed.

This commit only affects ECDH benchmarking, ECDSA still uses the low-level API.

### CODE REVIEW

File: apps/speed.c [hunk 1]
```diff
@@ -1040,33 +1038,31 @@ static int ECDSA_verify_loop(void *args)
...
+                            size_t *outlen,
+                            EVP_PKEY_CTX *ctx)
 {
-    loopargs_t *tempargs = *(loopargs_t **)args;
-    EC_KEY **ecdh_a = tempargs->ecdh_a;
-    EC_KEY **ecdh_b = tempargs->ecdh_b;
-    unsigned char *secret_a = tempargs->secret_a;
+    if( !EVP_PKEY_derive(ctx, derived_secret, outlen) ) {
+        /* FIXME: handle errors */
+        ;
```
Comment (levitte) [inline apps/speed.c]: If you're going to handle errors, it would be a good idea if this function returned `int` rather than `void`...

File: apps/speed.c [hunk 2]
```diff
@@ -2561,85 +2557,78 @@ int speed_main(int argc, char **argv)
...
+            EVP_PKEY_CTX *kctx = NULL, *ctx = NULL;
+            EVP_PKEY *key_A = NULL, *key_B = NULL;
+
+            if (testnum == R_EC_X25519) {
+                kctx = EVP_PKEY_CTX_new_id(test_curves[testnum], NULL); /* keygen ctx from NID */
+            } else {
+                EVP_PKEY_CTX *pctx = NULL;
+                EVP_PKEY *params = NULL;
+
+                if(     /* Create the context for parameter generation */
```
Comment (levitte) [inline apps/speed.c]: Style wise, we prefer a space between `if` and `(`.

File: apps/speed.c [hunk 3]
```diff
@@ -2561,85 +2557,78 @@ int speed_main(int argc, char **argv)
...
-            ERR_print_errors(bio_err);
-            rsa_count = 1;
-        } else {
-            for (i = 0; i < loopargs_len; i++) {
-                /* generate two ECDH key pairs */
-                if (!EC_KEY_generate_key(loopargs[i].ecdh_a[testnum]) ||
-                        !EC_KEY_generate_key(loopargs[i].ecdh_b[testnum])) {
-                    BIO_printf(bio_err, "ECDH key generation failure.\n");
-                    ERR_print_errors(bio_err);
+            EVP_PKEY_CTX *kctx = NULL, *ctx = NULL;
```
Comment (stevelinsell) [inline apps/speed.c]: You'll probably need to change all the lines that perform multiple declarations or operations on one line to be on separate lines in order to meet the OpenSSL coding style. This affects lines 2562, 2563, 2589, 2590, 2617.

File: apps/speed.c [hunk 4]
```diff
@@ -2561,85 +2557,78 @@ int speed_main(int argc, char **argv)
...
+
+                if(     /* Create the context for parameter generation */
+                        !(pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL)) ||
+                        /* Initialise the parameter generation */
+                        !EVP_PKEY_paramgen_init(pctx) ||
+                        /* Set the curve by NID */
+                        !EVP_PKEY_CTX_set_ec_paramgen_curve_nid(pctx, test_curves[testnum]) ||
+                        /* Create the parameter object params */
+                        !EVP_PKEY_paramgen(pctx, &params) ||
+                        0) {
```
Comment (stevelinsell) [inline apps/speed.c]: I'm not familiar with the '|| 0' clauses at the end of your if statements. What does this add over and above a normal if statement without them? Is it just about layout if you want to add another clause?

File: apps/speed.c [hunk 5]
```diff
@@ -2561,85 +2557,78 @@ int speed_main(int argc, char **argv)
...
+                BIO_printf(bio_err, "ECDH key generation failure.\n");
+                ERR_print_errors(bio_err);
+                rsa_count = 1;
+                break;
+            }
+
+            loopargs[i].ecdh_ctx[testnum] = ctx;
+
+            EVP_PKEY_CTX_free(kctx); kctx = NULL;
+        }
```
Comment (stevelinsell) [inline apps/speed.c]: Traditionally in speed you would have a test here to perform 1 operation. If the test passes then you go ahead with the loop of timed tests otherwise you fail. The actually timing loop is then not required to check for errors (for maximum efficiency) and is assumed to be running correctly. This isn't done consistently throughout speed as different people have done different things as they have added new tests. Previously for ECDH there were 2 ECDH_compute_key operations performed and compared for length as test before running the timed loop. This appears to have not been converted in the new code, was this intentional?

File: apps/speed.c [hunk 6]
```diff
@@ -2561,85 +2557,78 @@ int speed_main(int argc, char **argv)
...
-                    int secret_size_a, secret_size_b;
-                    /*
-                     * If field size is not more than 24 octets, then use SHA-1
-                     * hash of result; otherwise, use result (see section 4.8 of
-                     * draft-ietf-tls-ecc-03.txt).
-                     */
-                    int field_size = EC_GROUP_get_degree(
-                            EC_KEY_get0_group(loopargs[i].ecdh_a[testnum]));
-
-                    if (field_size <= 24 * 8) {                 /* 192 bits */
```
Comment (stevelinsell) [inline apps/speed.c]: In the old code when you ran ECDH tests there was an option to either run with or without a kdf operation. Using the EVP_PKEY interfaces the kdf operation is now not performed (it would need to be a separate step). Is this intentional, so that going forward it no longer makes sense to be able to time ECDH operations that take account of the kdf operation?

File: apps/speed.c [hunk 7]
```diff
@@ -1040,33 +1038,31 @@ static int ECDSA_verify_loop(void *args)
 /* ******************************************************************** */
 static long ecdh_c[EC_NUM][1];
 
-static int ECDH_compute_key_loop(void *args)
+static void ECDH_EVP_derive_key(unsigned char *derived_secret,
```
Comment (stevelinsell) [inline apps/speed.c]: Does writing this wrapper function buy anything over and above making the call directly at line 1061? At the moment it doesn't seem to add anything?

Comment (romen) [inline apps/speed.c]: Actually it does not, I was trying to mimic the structure of the old code as close as possible, do you think it would be better to just squeeze the code in just one function?

File: apps/speed.c (hunk 3 above)
Comment (romen) [inline apps/speed.c]: There are some coding style changes still pending that are included in my branch but still don't show up here (as a consequence of the github outage some hours ago).
I'll make sure to verify that this convention is applied! Thanks!

File: apps/speed.c (hunk 4 above)
Comment (romen) [inline apps/speed.c]: Yes, it's just for that, to be able to add/remove clauses without affecting other lines of code every time. The compiler is going to strip that away usually, so it should just be a matter of personal taste: would it be better if I removed that clause?

File: apps/speed.c (hunk 5 above)
Comment (romen) [inline apps/speed.c]: I addressed this issue in 2da6cc827e444eb142040f68527545db6e545245

File: apps/speed.c (hunk 4 above)
Comment (romen) [inline apps/speed.c]: I ended up removing the "|| 0" clause.

File: apps/speed.c (hunk 7 above)
Comment (romen) [inline apps/speed.c]: This was changed in 3f435aa95820540daf4366f3da8f1c667523b31b.

Now we only have the ECDH_EVP_derive_key_loop function that directly calls the EVP_PKEY_derive function without any intermediate wrapper.

File: apps/speed.c (hunk 1 above)
Comment (romen) [inline apps/speed.c]: Thanks for the feedback. I addressed this proposed change in various commits. The most relevant one, thanks to @stevelinsell  feedback, is 2da6cc827e444eb142040f68527545db6e545245.

File: apps/speed.c (hunk 2 above)
Comment (romen) [inline apps/speed.c]: This and other coding style changes were fixed, both manually and using the `util/openssl-format-source` utility: fb155b35495043fd1e479c9344f87cfe929c0658, c8f1e4f64c3c90258195642c5072acce94f7d422, eb35ac034306181cc024eddb4bf1a8d518e7eb08.

File: apps/speed.c (hunk 6 above)
Comment (romen) [inline apps/speed.c]: I'm not sure how to address this one: being completely honest I believed that the EVP interface, being at an higher abstraction level, was already taking care of using the right KDF when needed.

Digging into this I noticed an [old thread on the openssl-dev ML](https://mta.openssl.org/pipermail/openssl-users/2015-June/001661.html) but it doesn't seem to be conclusive on the right way to add a KDF step inside the EVP_PKEY_derive behaviour.

So we might discuss more on:
-  do we need to enforce the KDF step inside the benchmark loop? why?
-  should we try to make it happen inside `EVP_PKEY_derive()` using control fields in the `EVP_PKEY_CTX` structure? or should we perform it as a separate step inside the benchmark loop?
- assuming we will require KDF, when should we enforce it? the old code referenced draft-ietf-tls-ecc-03; should we still honor this? is there a "right" way to get the field size using the EVP API?
```
                   /*
                     * If field size is not more than 24 octets, then use SHA-1
                     * hash of result; otherwise, use result (see section 4.8 of
                     * draft-ietf-tls-ecc-03.txt).
                     */
```

**Comment by richsalz** [comment]
I just want to say that I am very excited to see this community collaboration here!

File: apps/speed.c (hunk 6 above)
Comment (stevelinsell) [inline apps/speed.c]: I have looked at this a little more. The EVP_PKEY_CTX_set_ecdh_* calls can only be called when we are using a CTX containing a EVP_PKEY_EC, i.e. all elliptical curves apart from X25519. Looking at the code in [crypto/ec/ecx_meth.c](https://github.com/openssl/openssl/blob/master/crypto/ec/ecx_meth.c) the EVP_PKEY_METHOD for X25519 does not contain ctrl message support for adding a KDF nor is there anywhere in the pkey_ecx_derive function where one is used, although my understanding is that a KDF can be used with X25519. As such I think we have to exclude X25519 from setting the KDF. Maybe I'm just missing something obvious, anyone else care to comment on this?
In terms of the other (non X25519) curves, I think you can get the field_size for those by using the:
EVP_PKEY_get1_EC_KEY(...)
function call to get an EC_KEY pointer and then use the same functions:
EC_KEY_get0_group(...)
EC_GROUP_get_degree(...)
as used in the existing speed implementation to get the actual field size.

Comment (romen) [inline apps/speed.c]: Well, X25519 output is over the 24B threshold so the KDF step shouldn't be enforced anyway. 

My point is: why do we need to keep this in the benchmarked loop?  It is my understanding that TLS is enforcing the KDF step at an higher level altogether (I'm still trying to verify this with s_client and s_server but having an hard time trying to enforce a 192 bit or lower curve for ECDH/ECDHE).

Also the section that mandated KDF step on the draft standard is not currently enforcing the same criteria so we need to put more thought on it.

What do you think?

Comment (stevelinsell) [inline apps/speed.c]: To be honest I have no idea why the KDF operation is included in the measurements, I can only think that historically someone thought the numbers made more sense to include both operations (perhaps to be more realistic when calculating the cost of a handshake) Personally I have no issue with whether it is done or not but someone in the community maybe expecting the KDF operation to be included, I was only suggesting we keep it because this is really a bug fix so I thought we should try and keep behaviour as close as possible to how it was. I'm easy though, any feedback from any other interested parties would be appreciated.

If you want to test s_server/s_client with a lower bit curve then you'll need to change some of the build flags (I think?) as by default OpenSSL removed default support for NIST curves smaller than 256 bits at the libSSL level. I can't remember offhand exactly what you need to set.

In a way I'm not sure it even makes sense to be measuring the speed of NIST curves less than 256 bit now, seeing as we aren't recommended to use them going forward anyway, With the current KDF speed behaviour that would mean all the curves supported by default would not need to have a KDF applied. That would solve the issue.

**Comment by stevelinsell** [comment]
I've downloaded and run some numbers this morning (unfortunately without your last commit for the memcmp).
These are run on an x86_64 platform with an Intel Xeon E5-2658 v3 cpu @ 2.20Ghz

  ```
                          Before op/s  After op/s  Diff %
 160 bit ecdh (secp160r1)     3667.1       3658.3     -0.24
 192 bit ecdh (nistp192)      3057.6       3105.5      1.57
 224 bit ecdh (nistp224)      2250.3       2256.2      0.26
 256 bit ecdh (nistp256)      9196.5       9196.0     -0.01
 384 bit ecdh (nistp384)       975.5        994.3      1.93
 521 bit ecdh (nistp521)       479.5        481.4      0.40
 163 bit ecdh (nistk163)      3312.0       3338.8      0.81
 233 bit ecdh (nistk233)      2613.6       2651.2      1.44
 283 bit ecdh (nistk283)      1456.9       1453.7     -0.22
 409 bit ecdh (nistk409)       912.3        918.1      0.64
 571 bit ecdh (nistk571)       385.1        393.7      2.23
 163 bit ecdh (nistb163)      3234.4       3168.0     -2.05
 233 bit ecdh (nistb233)      2561.0       2513.9     -1.84
 283 bit ecdh (nistb283)      1367.4       1362.8     -0.34
 409 bit ecdh (nistb409)       862.6        856.3     -0.73
 571 bit ecdh (nistb571)       361.8        361.3     -0.14
 253 bit ecdh (X25519)           inf       5980.9       N/A
```

To be honest the difference between running at the EVP level and the lower level is so small as to be hidden by the natural variation between runs, so nothing can really be read into the numbers above. I was interested that even though there is no KDF now it didn't seem to make a significant difference to the numbers for the small curves, I guess the KDF calculation is so small compared to the EC operation that again I think the natural variation hides any differences. 
Anyway I think we can say that this change is unlikely to make much difference in the numbers people measure.

I also ran your commits when using a Hardware Acceleration engine both synchronously and asynchronously and can confirm that both function as expected.

**Comment by romen** [comment]
I also went ahead and made sure that during TLS/SSL handshake, even when you set a 192 bit curve for ECDHE, `ECDH_compute_key()` is not passed a KDF function pointer, hence KDF is definitely done at an higher level if needed.

## Huge wall of text, TL;DR at the end

This is how I tested it (also for future reference as it wasn't easy to find a way to test this):
-  `openssl req -x509 -sha256 -nodes -days 365 -newkey rsa:2048 -keyout /tmp/privateKey.pem  -out /tmp/cert.pem`

  to generate an RSA (or whatever) privatekey and self-signed certificate

- `cat /tmp/cert.pem /tmp/privateKey.pem > /tmp/server.pem`

  to concatenate private key and certificate in a single file
-  `gdb --args openssl s_server -cert /tmp/server.pem -state -msg -cipher $(apps/openssl ciphers ECDHE) -curves prime192v1`

  ```
GNU gdb (Ubuntu 7.11.1-0ubuntu1~16.04) 7.11.1
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from apps/openssl...done.
(gdb) br ECDH_compute_key
Function "ECDH_compute_key" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (ECDH_compute_key) pending.
(gdb) run

[...] // concurrently run:
openssl s_client -cipher $(apps/openssl ciphers ECDHE) -curves prime192v1 < /dev/null
[...]


Breakpoint 1, ECDH_compute_key (out=0x6e0990, outlen=24, pub_key=0x6e03f0, eckey=0x6e01f0, KDF=0x0) at crypto/ec/ec_kmeth.c:131
131     {
(gdb) print KDF
$1 = (void *(*)(const void *, size_t, void *, size_t *)) 0x0
(gdb) print eckey->group->curve_name 
$2 = 409
(gdb) continue
```
  * so `KDF` is `NULL`

  * to check that the right curve is being used let's verify the NID matches: `egrep "409$" include/openssl/obj_mac.h`

     ```
#define NID_X9_62_prime192v1            409
```

-  and here is the output from the concurrent client: `openssl s_client -cipher $(apps/openssl ciphers ECDHE) -curves prime192v1 < /dev/null` from which we can easily verify that a 192bit curve is being used for ECDHE and that the handshake is correctly completed according to the TLS 1.2 standard!

  ```
CONNECTED(00000003)
depth=0 C = AU, ST = Some-State, O = Internet Widgits Pty Ltd
verify error:num=18:self signed certificate
verify return:1
depth=0 C = AU, ST = Some-State, O = Internet Widgits Pty Ltd
verify return:1
---
Certificate chain
 0 s:/C=AU/ST=Some-State/O=Internet Widgits Pty Ltd
   i:/C=AU/ST=Some-State/O=Internet Widgits Pty Ltd
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIDXTCCAkWgAwIBAgIJAJySXOXmaTW+MA0GCSqGSIb3DQEBCwUAMEUxCzAJBgNV
BAYTAkFVMRMwEQYDVQQIDApTb21lLVN0YXRlMSEwHwYDVQQKDBhJbnRlcm5ldCBX
aWRnaXRzIFB0eSBMdGQwHhcNMTYxMDA2MTUyOTE4WhcNMTcxMDA2MTUyOTE4WjBF
MQswCQYDVQQGEwJBVTETMBEGA1UECAwKU29tZS1TdGF0ZTEhMB8GA1UECgwYSW50
ZXJuZXQgV2lkZ2l0cyBQdHkgTHRkMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIB
CgKCAQEArXzARfMgxf+ofv498Md2HBTPOFwknT3UYXpm3GYaUEqLamNVVqPEP1kF
MaVXH6i5oCUcOn7HrGTV4DNORud3GJBP1NIWEiqEux74vTg4etn1l9QXOnjHd1O5
rCknUyVXByawfh0PhX3a52KQT9iy+p+/PNbheoPpn603iEr+wl9tHS1UbbtgU2QC
6x/w5RrZCs7td/DyWVRVwe9eDzEa7/ofJKzhs+aDeveSWSdUJEoeUNZG+zdBF/Jz
mzgxv+ZW7XqRqLj1zIQxvK6OOuBWdcpfGs8D3UA6f+8YeeJSWaGIDq0RtjAGPuZV
suAHMdfGB7gKhVlPxyiQ37XrtI+6pQIDAQABo1AwTjAdBgNVHQ4EFgQUacTZmJoV
p5PuQkT3uXqdFccW7XkwHwYDVR0jBBgwFoAUacTZmJoVp5PuQkT3uXqdFccW7Xkw
DAYDVR0TBAUwAwEB/zANBgkqhkiG9w0BAQsFAAOCAQEAImAK9dDDmxtLW0+qWSB7
IDQpMMNGILKFJKBGLD7nU0D2P3QMYOhX7PvdOCLtk1WtKz+uxcvIgaPfKozwTWaz
rMQAdD6MThuX32Yr5q8YKL7otMy/PUn1oG4OcO7PlFnnxsq1GJiv1VP7uHiUVKdq
BCRG1iKcEEH6L+CgTO4u/VHBlfys4OERdjF9x6Kvd3cMvKJNTp69ZxnsXzjU3KcM
NjYCBpfAnW4Cuq7ay73vwhKmxPCnbKM5yelcf4lF/QsbDYr8xwXz1dammhjajCW/
Ifj69eZt5EDdYdUbVAhRM7OQodA3uSamQ68yhz6rwG8TqgmVV4gfa1FTm7NMTAlL
3A==
-----END CERTIFICATE-----
subject=/C=AU/ST=Some-State/O=Internet Widgits Pty Ltd
issuer=/C=AU/ST=Some-State/O=Internet Widgits Pty Ltd
---
No client certificate CA names sent
Peer signing digest: SHA512
Server Temp Key: ECDH, P-192, 192 bits
---
SSL handshake has read 1507 bytes and written 270 bytes
Verification error: self signed certificate
---
New, TLSv1.2, Cipher is ECDHE-RSA-AES256-GCM-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-GCM-SHA384
    Session-ID: 9885108F851947F0EE76599F31C7A501155F0824D7F523AD679C06E34CA5C797
    Session-ID-ctx: 
    Master-Key: E1B36845E4C3185A865CEEB105BE36F252746DF70B090B4357CF0BE40788C1565AA73EBFCBA1212CAB7D8EC5928C9D5D
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 7200 (seconds)
    TLS session ticket:
    0000 - a5 ee ee 0c 17 56 7c 7f-82 0a 42 82 bf 96 0c 36   .....V|...B....6
    0010 - bc 00 ed 79 2a d3 57 b7-05 2e 11 02 bc 8e ad 0b   ...y*.W.........
    0020 - 0a a3 76 3b 2f a2 7b d2-6e 0a 9b 3a b5 cc 81 37   ..v;/.{.n..:...7
    0030 - fc 87 5d d1 19 cf 57 ad-eb ab 70 8d 93 54 82 e0   ..]...W...p..T..
    0040 - a1 6c b2 2c 83 9d 4d 94-c0 aa a7 0b f3 73 43 f4   .l.,..M......sC.
    0050 - e1 fd a1 fe d9 3e f9 8b-d2 3b 33 fd 7d 58 59 d6   .....>...;3.}XY.
    0060 - 11 46 c5 9f 24 f5 87 e3-ee f7 e6 45 84 e3 b3 79   .F..$......E...y
    0070 - 8e 34 ca ec 92 94 cf 75-53 d7 9e 3b 47 08 e0 f4   .4.....uS..;G...
    0080 - c4 15 18 f4 7a 56 9a 09-7c 85 99 7b 21 c3 ed fa   ....zV..|..{!...
    0090 - 67 ad 34 31 12 8e 36 e1-05 f8 5d c2 9f 36 e7 c2   g.41..6...]..6..

    Start Time: 1475768956
    Timeout   : 7200 (sec)
    Verify return code: 18 (self signed certificate)
    Extended master secret: yes
---
DONE
```




## TL;DR

tests confirm that even when the curve field elements are 24-bytes or less, the current (1.1.0b) version of OpenSSL does not execute the KDF step at the `ECDH_compute_key()` level anyway, so I really don't understand why the current `apps/speed.c` code is enforcing it and don't see why we should keep it this way switching to the `EVP` API.

**Comment by romen** [comment]
Can anyone suggest me how to properly address the `coverage/coveralls` tests that keep failing?

**Comment by stevelinsell** [comment]
In Terms of the KDF. Your investigation makes sense. RFC 4492 makes it clear that the KDF at the TLS Layer is achieved by setting the x co-ordinate of the ECDH elliptic curve shared secret to the pre-master secret. The pre-master secret is generated by a hash operation via the PRF operation. In effect you are relying on the PRF operation to provide the hashing functionality the KDF would have done, and as you state this is done outside of the whole API we are dealing with.
I'm guessing that the KDF support within ECDH_compute_key is meant to be utilized by applications that are using ECDH primatives but are not using TLS on top.  
Do we need to support that use case, I'm not sure. I reckon it's best not to add it and if anyone is interested in that support they can add a pull request to add that functionality on top.

**Comment by romen** [comment]
@stevelinsell I agree, I would leave it out for now and wait for someone that really has a use for this use case.

Latest commits remove some spurious references to the KDF step, avoid treating X25519 as a special case, and fix a minor bug introduced by the latter that was cluttering the error queue.

I still need to address the failing coverage tests, but need guidance on how to do so.

**Comment by stevelinsell** [comment]
Great, I'll give your commits a whirl.
I wouldn't worry too much about the code coverage. If you click on the details I think it's just telling you that you've added more lines of code than you've removed, effectively increasing the amount of lines of code in the project. Because speed.c does not have any coverage tests at all this decreases the overall coverage of the project by 0.02% raising the red flag. Unless we were to implement code coverage testing for speed then we can't really do anything about this. I don't think this will be a barrier to the pull request being accepted.

**Comment by richsalz** [comment]
Great work.  Now we just need a CLA from you, @romen

**Comment by romen** [comment]
@richsalz CLA just sent!

**Comment by richsalz** [comment]
Great, now one last request:  please rebase on a recent master....  Then I wil merge ASAP.

**Comment by romen** [comment]
@richsalz done, just waiting for the CI checks to pass! ;)

**Comment by richsalz** [comment]
Thanks very much, guys.  Nice to see the "joint development effort" :)  Merged to master.
